src/appengine/handlers/cron/fuzz_strategy_selection.py

```diff
@@ -0,0 +1,135 @@
+    SUM(EXP(strategy_avg_edges / temperature)) OVER() AS exp_sum,
+    strategy,
+    strategy_count
+  FROM
+    (SELECT
+      /* Standardize the new edges data and take averages per strategy. */
+      AVG((new_edges - overall_avg_new_edges) / overall_stddev_new_edges) AS strategy_avg_edges,
+      strategy,
+      /* Change temperature parameter here. */
+      .5 as temperature,
+      COUNT(*) AS strategy_count
+    FROM
+      (SELECT
+        fuzzer,
+        CONCAT(s_radamsa, s_max_len, s_ml_rnn, s_vp, s_corpus, s_fork, s_subset, s_recommended_dict) AS strategy,
+        fuzzer_stddev,
+        AVG(new_edges) OVER() AS overall_avg_new_edges,
+        STDDEV(new_edges) OVER() AS overall_stddev_new_edges,
+        new_edges
+      FROM
+        (SELECT
+          fuzzer,
+          IF(strategy_corpus_mutations_radamsa > 0, "radamsa,", "") AS s_radamsa,
+          IF(strategy_random_max_len > 0, "max len,", "") AS s_max_len,
+          IF(strategy_corpus_mutations_ml_rnn > 0,"ml rnn,", "") AS s_ml_rnn, 
+          IF(strategy_value_profile > 0, "value profile,", "") AS s_vp, 
+          IF(strategy_corpus_mutations > 0, "corpus,", "") AS s_corpus,
+          IF(strategy_fork > 0, "fork,", "") AS s_fork,
+          IF(strategy_corpus_subset > 0, "subset,", "") AS s_subset,
+          IF(strategy_recommended_dict > 0, "dict,", "") AS s_recommended_dict,
+          STDDEV(new_edges) OVER(PARTITION by fuzzer) AS fuzzer_stddev,
+          new_edges
+        FROM 
+          libFuzzer_stats.TestcaseRun
+        WHERE
+           ((strategy_mutator_plugin = 0) OR (strategy_mutator_plugin IS NULL)) AND
+            DATE_DIFF(cast(current_timestamp() AS DATE), cast(_PARTITIONTIME AS DATE), DAY) < 31 AND
+            ((strategy_corpus_mutations = 0) OR (strategy_corpus_mutations IS NULL)) AND
+            ((strategy_handle_unstable = 0) OR (strategy_handle_unstable IS NULL)) AND
+            ((strategy_weighted_mutations = 0) OR (strategy_weighted_mutations IS NULL)))
+      WHERE
+        /* Filter for unstable targets. */
+        fuzzer_stddev < 150)
+    GROUP BY
+      strategy))
+ORDER BY
+  bandit_weight DESC
+"""
+
+
+def _query_multi_armed_bandit_probs(client):
+  """Get query results.
+
+  Queries above BANDIT_PROBABILITY_QUERY and yields results
+  from bigquery. This query is sorted by strategies implemented."""
+  return client.query(query=BANDIT_PROBABILITY_QUERY)
+
+
+def _upload_fuzz_strategy_weights(client):
+  """Uploads queried data into datastore.
+
+  Upload query results to datastore to use as new probabilities.
+  Probabilities are based on new_edges feature."""
+  strategy_data = []
+  data = _query_multi_armed_bandit_probs(client)
+
+  for row in data:
+    curr_strategy = data_types.FuzzStrategyProbability()
+    curr_strategy.strategy_name = str(row['strategy'])
+    curr_strategy.probability = float(row['bandit_weight'])
+    strategy_data.append(curr_strategy)
+  ndb.delete_multi([
```

Comment:
And this?

Reply:
There is not any mention of this function throwing exceptions in the documentation, so I think this should be fine

Reply:
We've seen a lot of intermittent failures but I actually prefer the current behavior in the event of an exception (failing to add the new entries). The code that uses this wouldn't be able to handle duplicate entries anyway, so that would put us in a bad state.

Reply:
Thanks for clarifying!